controllers/apps/transformer_cluster_service.go

```diff
@@ -167,22 +174,67 @@ func (t *clusterServiceTransformer) buildService(transCtx *clusterTransformConte
 		Optimize4ExternalTraffic()
 
 	if len(clusterService.ComponentSelector) > 0 {
+		builder.AddSelector(constant.KBAppComponentLabelKey, clusterService.ComponentSelector)
+	}
+
+	if len(clusterService.ShardSelector) > 0 {
+		builder.AddSelector(constant.KBAppShardTemplateNameLabelKey, clusterService.ShardSelector)
+	}
+
+	if len(clusterService.RoleSelector) > 0 {
 		compDef, err := t.checkComponent(transCtx, clusterService)
 		if err != nil {
 			return nil, err
 		}
-		builder.AddSelector(constant.KBAppComponentLabelKey, clusterService.ComponentSelector)
-
-		if len(clusterService.RoleSelector) > 0 {
-			if err := t.checkComponentRoles(compDef, clusterService); err != nil {
-				return nil, err
-			}
-			builder.AddSelector(constant.RoleLabelKey, clusterService.RoleSelector)
+		if err := t.checkComponentRoles(compDef, clusterService); err != nil {
+			return nil, err
 		}
+		builder.AddSelector(constant.RoleLabelKey, clusterService.RoleSelector)
 	}
+
 	return builder.GetObject(), nil
 }
 
+func (t *clusterServiceTransformer) genMultiServiceIfNeed(cluster *appsv1alpha1.Cluster, clusterService *appsv1alpha1.ClusterService) ([]*appsv1alpha1.ClusterService, error) {
+	if len(clusterService.ShardSelector) == 0 || len(cluster.Spec.ShardingSpecs) == 0 {
+		return []*appsv1alpha1.ClusterService{clusterService}, nil
+	}
+
+	shardingTemplateName := ""
+	shards := int32(0)
+	for _, shardingSpec := range cluster.Spec.ShardingSpecs {
+		if shardingSpec.Name != clusterService.ShardSelector {
+			continue
+		}
+		shardingTemplateName = shardingSpec.Name
+		shards = shardingSpec.Shards
+	}
+
+	if len(shardingTemplateName) == 0 {
+		return nil, fmt.Errorf("the shardSelector of service is not defined, service: %s, shard: %s", clusterService.Name, clusterService.ShardSelector)
+	}
+
+	enableShardOrdinalSvcTplNameList, ok := cluster.Annotations[constant.ShardOrdinalSvcAnnotationKey]
```

Comment:
it's service for shard, nothing about ordinal

Reply:
> it's service for shard, nothing about ordinal

done